File: cpp/src/geometry.cpp
#include "rgbd/geometry.hpp"

#include <Eigen/LU>
#include <Eigen/SVD>

#include <algorithm>
#include <cmath>
#include <cstdint>
#include <limits>
#include <map>
#include <numeric>
#include <stdexcept>
#include <tuple>
#include <utility>

namespace rgbd {
namespace {

class KdTree3d {
public:
    explicit KdTree3d(const PointCloud& points) : points_(points), indices_(points.size()) {
        std::iota(indices_.begin(), indices_.end(), 0U);
        nodes_.reserve(points.size());
        root_ = build(0, indices_.size(), 0);
    }

    std::pair<std::size_t, double> nearest(const Vec3& query) const {
        std::size_t best_index = 0;
        double best_squared = std::numeric_limits<double>::infinity();
        search(root_, query, best_index, best_squared);
        return {best_index, best_squared};
    }

private:
    struct Node {
        std::size_t point_index = 0;
        int left = -1;
        int right = -1;
        int axis = 0;
    };

    int build(std::size_t begin, std::size_t end, int depth) {
        if (begin >= end) {
            return -1;
        }
        const int axis = depth % 3;
        const std::size_t middle = begin + (end - begin) / 2;
        std::nth_element(
            indices_.begin() + static_cast<std::ptrdiff_t>(begin),
            indices_.begin() + static_cast<std::ptrdiff_t>(middle),
            indices_.begin() + static_cast<std::ptrdiff_t>(end),
            [&](std::size_t lhs, std::size_t rhs) {
                if (points_[lhs][axis] == points_[rhs][axis]) {
                    return lhs < rhs;
                }
                return points_[lhs][axis] < points_[rhs][axis];
            });
        const int node_index = static_cast<int>(nodes_.size());
        nodes_.push_back({indices_[middle], -1, -1, axis});
        const int left = build(begin, middle, depth + 1);
        const int right = build(middle + 1, end, depth + 1);
        nodes_[static_cast<std::size_t>(node_index)].left = left;
        nodes_[static_cast<std::size_t>(node_index)].right = right;
        return node_index;
    }

    void search(int node_index, const Vec3& query, std::size_t& best_index, double& best_squared) const {
        if (node_index < 0) {
            return;
        }
        const Node& node = nodes_[static_cast<std::size_t>(node_index)];
        const Vec3& point = points_[node.point_index];
        const double squared = (query - point).squaredNorm();
        if (squared < best_squared || (squared == best_squared && node.point_index < best_index)) {
            best_squared = squared;
            best_index = node.point_index;
        }
        const double difference = query[node.axis] - point[node.axis];
        const int near_child = difference <= 0.0 ? node.left : node.right;
        const int far_child = difference <= 0.0 ? node.right : node.left;
        search(near_child, query, best_index, best_squared);
        if (difference * difference <= best_squared) {
            search(far_child, query, best_index, best_squared);
        }
    }

    const PointCloud& points_;
    std::vector<std::size_t> indices_;
    std::vector<Node> nodes_;
    int root_ = -1;
};

struct Matches {
    PointCloud source;
    PointCloud target;
    double all_point_rmse_m = 0.0;
};

Matches find_matches(
    const PointCloud& source,
    const PointCloud& target,
    const KdTree3d& tree,
    double max_distance) {
    Matches matches;
    matches.source.reserve(source.size());
    matches.target.reserve(source.size());
    const double max_squared = max_distance * max_distance;
    double all_squared_sum = 0.0;
    for (const Vec3& point : source) {
        const auto [index, squared] = tree.nearest(point);
        all_squared_sum += squared;
        if (squared <= max_squared) {
            matches.source.push_back(point);
            matches.target.push_back(target[index]);
        }
    }
    matches.all_point_rmse_m = std::sqrt(all_squared_sum / static_cast<double>(source.size()));
    return matches;
}

}  // namespace

PointCloud transform_points(const PointCloud& points, const RigidTransform& transform) {
    PointCloud transformed;
    transformed.reserve(points.size());
    for (const Vec3& point : points) {
        transformed.push_back(transform.rotation * point + transform.translation);
    }
    return transformed;
}

RigidTransform kabsch(const PointCloud& source, const PointCloud& target) {
    if (source.size() != target.size() || source.size() < 3) {
        throw std::invalid_argument("Kabsch requires equal point clouds with at least three points");
    }
    Vec3 source_center = Vec3::Zero();
    Vec3 target_center = Vec3::Zero();
    for (std::size_t i = 0; i < source.size(); ++i) {
        source_center += source[i];
        target_center += target[i];
    }
    source_center /= static_cast<double>(source.size());
    target_center /= static_cast<double>(target.size());

    Mat3 covariance = Mat3::Zero();
    for (std::size_t i = 0; i < source.size(); ++i) {
        covariance += (source[i] - source_center) * (target[i] - target_center).transpose();
    }
    const Eigen::JacobiSVD<Mat3> svd(covariance, Eigen::ComputeFullU | Eigen::ComputeFullV);
    Mat3 correction = Mat3::Identity();
    if ((svd.matrixV() * svd.matrixU().transpose()).determinant() < 0.0) {
        correction(2, 2) = -1.0;
    }
    RigidTransform result;
    result.rotation = svd.matrixV() * correction * svd.matrixU().transpose();
    result.translation = target_center - result.rotation * source_center;
    return result;
}

PointCloud voxel_downsample(const PointCloud& points, double voxel_size_m) {
    if (points.size() < 3 || voxel_size_m <= 0.0) {
        throw std::invalid_argument("voxel_downsample requires at least three points and a positive voxel size");
    }
    struct Accumulator {
        Vec3 sum = Vec3::Zero();
        std::size_t count = 0;
    };
    std::map<std::tuple<std::int64_t, std::int64_t, std::int64_t>, Accumulator> voxels;
    for (const Vec3& point : points) {
        const auto key = std::make_tuple(
            static_cast<std::int64_t>(std::floor(point.x() / voxel_size_m)),
            static_cast<std::int64_t>(std::floor(point.y() / voxel_size_m)),
            static_cast<std::int64_t>(std::floor(point.z() / voxel_size_m)));
        auto& accumulator = voxels[key];
        accumulator.sum += point;
        ++accumulator.count;
    }
    PointCloud output;
    output.reserve(voxels.size());
    for (const auto& [key, accumulator] : voxels) {
        static_cast<void>(key);
        output.push_back(accumulator.sum / static_cast<double>(accumulator.count));
    }
    return output;
}

IcpResult icp_point_to_point(
    const PointCloud& source,
    const PointCloud& target,
    int max_iterations,
    double tolerance,
    double max_correspondence_m) {
    if (source.size() < 3 || target.size() < 3 || max_iterations < 1 || tolerance < 0.0 || max_correspondence_m <= 0.0) {
        throw std::invalid_argument("invalid ICP inputs or parameters");
    }
    const KdTree3d tree(target);
    PointCloud aligned = source;
    RigidTransform total;
    double previous_rmse = std::numeric_limits<double>::infinity();
    IcpResult result;

    for (int iteration = 1; iteration <= max_iterations; ++iteration) {
        const Matches matches = find_matches(aligned, target, tree, max_correspondence_m);
        if (matches.source.size() < 3) {
            throw std::runtime_error("fewer than three valid ICP correspondences");
        }
        const RigidTransform delta = kabsch(matches.source, matches.target);
        aligned = transform_points(aligned, delta);
        total.translation = delta.rotation * total.translation + delta.translation;
        total.rotation = delta.rotation * total.rotation;

        const Matches updated = find_matches(aligned, target, tree, max_correspondence_m);
        result = {total, updated.all_point_rmse_m, iteration, matches.source.size()};
        if (std::abs(previous_rmse - updated.all_point_rmse_m) < tolerance) {
            return result;
        }
        previous_rmse = updated.all_point_rmse_m;
    }
    return result;
}

}  // namespace rgbd

File: src/geometry.py
"""不依赖 Open3D 的最小点云几何与 ICP 实现。

这里的目标是学习算法和验证结果，不追求生产级性能。
"""

from __future__ import annotations

from dataclasses import dataclass

import numpy as np
from scipy.spatial import cKDTree


Array = np.ndarray


@dataclass
class ICPResult:
    rotation: Array
    translation: Array
    aligned_source: Array
    rmse: float
    iterations: int
    correspondences: int


def as_points(points: Array) -> Array:
    """校验并返回形状为 (N, 3) 的点云。"""
    points = np.asarray(points, dtype=float)
    if points.ndim != 2 or points.shape[1] != 3:
        raise ValueError(f"点云应为 (N, 3)，实际为 {points.shape}")
    if len(points) < 3:
        raise ValueError("至少需要三个点")
    return points


def transform_points(points: Array, rotation: Array, translation: Array) -> Array:
    """对行向量点云执行 p' = R p + t。"""
    points = as_points(points)
    rotation = np.asarray(rotation, dtype=float)
    translation = np.asarray(translation, dtype=float).reshape(3)
    if rotation.shape != (3, 3):
        raise ValueError("rotation 应为 (3, 3)")
    return (rotation @ points.T).T + translation


def kabsch(source: Array, target: Array) -> tuple[Array, Array]:
    """已知一一对应点时，最小二乘估计 source 到 target 的刚体变换。"""
    source = as_points(source)
    target = as_points(target)
    if source.shape != target.shape:
        raise ValueError("source 与 target 必须点数相同且一一对应")

    source_center = source.mean(axis=0)
    target_center = target.mean(axis=0)
    source_zero = source - source_center
    target_zero = target - target_center

    covariance = source_zero.T @ target_zero
    u, _, vt = np.linalg.svd(covariance)
    rotation = vt.T @ u.T

    # SVD 可能产生镜像反射；刚体变换必须保持右手坐标系。
    if np.linalg.det(rotation) < 0:
        vt[-1, :] *= -1
        rotation = vt.T @ u.T

    translation = target_center - rotation @ source_center
    return rotation, translation


def voxel_downsample(points: Array, voxel_size: float) -> Array:
    """按体素内点的平均值进行最小化下采样。"""
    points = as_points(points)
    if voxel_size <= 0:
        raise ValueError("voxel_size 必须大于 0")

    voxel_id = np.floor(points / voxel_size).astype(np.int64)
    _, inverse = np.unique(voxel_id, axis=0, return_inverse=True)
    sums = np.zeros((inverse.max() + 1, 3), dtype=float)
    counts = np.bincount(inverse)
    np.add.at(sums, inverse, points)
    return sums / counts[:, None]


def statistical_outlier_filter(points: Array, neighbors: int = 20, std_ratio: float = 2.0) -> Array:
    """用邻域平均距离过滤孤立点。

    这是一个教学版实现；大点云的工程版本应优先使用 Open3D/PCL。
    """
    points = as_points(points)
    if not 1 <= neighbors < len(points):
        raise ValueError("neighbors 必须介于 1 与点数减一之间")
    if std_ratio < 0:
        raise ValueError("std_ratio 不能为负")

    distances, _ = cKDTree(points).query(points, k=neighbors + 1)
    mean_distance = distances[:, 1:].mean(axis=1)
    threshold = mean_distance.mean() + std_ratio * mean_distance.std()
    return points[mean_distance <= threshold]


def icp_point_to_point(
    source: Array,
    target: Array,
    *,
    max_iterations: int = 80,
    tolerance: float = 1e-8,
    max_correspondence_distance: float | None = None,
) -> ICPResult:
    """使用最近邻匹配和 Kabsch 更新的点到点 ICP。

    ICP 只在初始位姿足够接近时可靠。真实场景中通常需要特征匹配或 RANSAC 提供初值。
    """
    source = as_points(source)
    target = as_points(target)
    tree = cKDTree(target)

    aligned = source.copy()
    total_rotation = np.eye(3)
    total_translation = np.zeros(3)
    previous_rmse = np.inf

    for iteration in range(1, max_iterations + 1):
        distances, nearest_index = tree.query(aligned, k=1)
        valid = np.ones(len(aligned), dtype=bool)
        if max_correspondence_distance is not None:
            valid = distances <= max_correspondence_distance
        if valid.sum() < 3:
            raise RuntimeError("有效对应点少于 3 个；请放宽阈值或改善初始位姿")

        matched_source = aligned[valid]
        matched_target = target[nearest_index[valid]]
        delta_rotation, delta_translation = kabsch(matched_source, matched_target)
        aligned = transform_points(aligned, delta_rotation, delta_translation)

        total_rotation = delta_rotation @ total_rotation
        total_translation = delta_rotation @ total_translation + delta_translation

        new_distances, _ = tree.query(aligned, k=1)
        rmse = float(np.sqrt(np.mean(new_distances**2)))
        if abs(previous_rmse - rmse) < tolerance:
            return ICPResult(
                total_rotation,
                total_translation,
                aligned,
                rmse,
                iteration,
                int(valid.sum()),
            )
        previous_rmse = rmse

    return ICPResult(
        total_rotation,
        total_translation,
        aligned,
        previous_rmse,
        max_iterations,
        int(valid.sum()),
    )
